### ISSUE TITLE
lsp: elmls code action "Expose Function" has no effect

### ISSUE THREAD

**Issue by Massolari** [neovim/neovim#15828]
### Neovim version (nvim -v)

0.5.1 Release

### Vim (not Nvim) behaves the same?

Vim doesn't has a LSP native client

### Operating system/version

macOS 11.4

### Terminal name/version

kitty 0.23.1

### $TERM environment variable

xterm-kitty

### Installation

homebrew

### How to reproduce the issue

In an Elm project, try to expose something using `vim.lsp.buf.code_action()`

### Expected behavior

The function/type/other gets exposed in the module declaration.

### Actual behavior

![image](https://user-images.githubusercontent.com/12272702/135124904-5d916eec-b8b2-46e8-bfda-94e887bcc4ca.png)

**Comment by mjlbach** [comment]
Can you check if this happens on master?

**Comment by Massolari** [comment]
Just built it! This happens on master too:
![image](https://user-images.githubusercontent.com/12272702/135131941-8b242837-f873-4106-a080-7fb93fc61956.png)

![image](https://user-images.githubusercontent.com/12272702/135131993-0ab374a1-2d32-45db-a5f5-e282db61776f.png)

**Comment by mjlbach** [comment]
Can you send a minimal repro including the elm file and a minimal init.lua/vim (you can use the one in lspconfig). I can check the logs. You can also do this with:

```lua
vim.lsp.set_log_level("debug")
require('vim.lsp.log').set_format_func(vim.inspect)
```

**Comment by Massolari** [comment]
Repo: https://github.com/Massolari/elm-code-action-error
Open the `A.elm` file and try to expose the `wantToExpose` function using code action.

**Comment by Massolari** [comment]
About the log, i've got this:

```
[INFO][2021-09-28 15:44:03] ...lsp/handlers.lua:423	"A code action was requested"
[DEBUG][2021-09-28 15:44:03] .../vim/lsp/rpc.lua:496	"rpc.receive"	{
  id = 2,
  jsonrpc = "2.0",
  result = { {
      data = {
        fixId = "remove_unused"
      },
      diagnostics = { {
          data = {
            code = "unused_top_level",
            uri = "file:///Users/douglasmassolari/elm-code-action-error/src/A.elm"
          },
          message = "Unused top level definition `wantToExpose`",
          range = {
            end = {
              character = 12,
              line = 9
            },
            start = {
              character = 0,
              line = 9
            }
          },
          severity = 2,
          source = "ElmLS",
          tags = { 1 }
        } },
      edit = {
        changes = {
          ["file:///Users/douglasmassolari/elm-code-action-error/src/A.elm"] = { {
              newText = "",
              range = {
                end = {
                  character = 45,
                  line = 10
                },
                start = {
                  character = 0,
                  line = 8
                }
              }
            } }
        }
      },
      isPreferred = false,
      kind = "quickfix",
      title = "Remove function `wantToExpose`"
    }, {
      data = {
        actionName = "expose_function",
        range = {
          end = {
            character = 12,
            line = 9
          },
          start = {
            character = 0,
            line = 9
          }
        },
        refactorName = "expose_unexpose",
        uri = "file:///Users/douglasmassolari/elm-code-action-error/src/A.elm"
      },
      isPreferred = true,
      kind = "refactor",
      title = "Expose Function"
    } }
}
```

**Comment by mjlbach** [comment]
Where exactly am I supposed to trigger the code action? I get no code actions available with range/regular coe actions on wantToExpose on line 9

**Comment by Massolari** [comment]
Without error?!

**Comment by Massolari** [comment]
Are you with @elm-tooling/elm-language-server@2.2.1?

**Comment by mjlbach** [comment]
```
 ~/.local/share/ov/pkg/isaac_sim-2021.1.1                                                                          isaac-sim
❯ elm-language-server --version
2.2.1

 ~/.local/share/ov/pkg/isaac_sim-2021.1.1                                                                          isaac-sim
❯ elm --version
0.19.1

 ~/.local/share/ov/pkg/isaac_sim-2021.1.1                                                                          isaac-sim
❯ node --version
v14.17.6
```

**Comment by mjlbach** [comment]
How did you build neovim? Did you build from source without installing? You may have not set VIMRUNTIME and be using your intsalled runtime files from 0.5.

**Comment by mjlbach** [comment]
I'm not sure the code action is succesful (I don't think it is) but it doesn't throw the error you are seeing.

**Comment by Massolari** [comment]
Yes, I built from source but not installed it, I'm running directly from the `build` folder. I'll try to install it.
To be successful, the code action should add `wantToExpose` to the list of exposed symbols in the top of the file. It should be `module A exposing (exposed, wantToExpose)`

**Comment by Massolari** [comment]
It "works". I mean, it didnt, but it's not throwing that error.
We're on same page now

**Comment by mjlbach** [comment]
Ok, great! It seems like this is missing a command so I'm wondering if it's supposed to be implemented client side... I'm not as familiar with the implementation details of code actions as @mfussenegger

**Comment by mfussenegger** [comment]
The action doesn't have a `edit` or `command` property,  so it becomes a no-op.

```
{
      data = {
        actionName = "expose_function",
        range = {
          end = {
            character = 12,
            line = 9
          },
          start = {
            character = 0,
            line = 9
          }
        },
        refactorName = "expose_unexpose",
        uri = "file:///Users/douglasmassolari/elm-code-action-error/src/A.elm"
      },
      isPreferred = true,
      kind = "refactor",
      title = "Expose Function"
    } 

```


It could be that the server is not honoring the client capabilities and requires codeAction/resolve support.  You could try https://github.com/neovim/neovim/pull/15818 to see if that's the case. 

If it isn't I'm also out of ideas how it's supposed to work.

**Comment by Massolari** [comment]
With the version of this PR it worked!

**Comment by mjlbach** [comment]
That was a change in neovim 0.6, unrelated to this PR

**Comment by mjlbach** [comment]
Closed by #15818
